samza-api/src/main/java/org/apache/samza/system/SystemAdmin.java

```diff
@@ -22,12 +22,24 @@
 import java.util.Map;
 import java.util.Set;
 
+
 /**
  * Helper interface attached to an underlying system to fetch information about
  * streams, partitions, offsets, etc. This interface is useful for providing
  * utility methods that Samza needs in order to interact with a system.
  */
 public interface SystemAdmin {
+
+  /**
+   * Connect to the underlying system and prepare to begin serving requests
+   */
+  default void start() {};
```

Comment:
Yes. It currently applies to Kafka. I added these two methods in SystemAdmin interface because it seems to make sense to have start() and stop() in general, similar to SystemProducer() and SystemConsumer(). If this sounds reasonable in the long term, and there is no short term concern since we provided an empty default method for existing users, maybe we could just add these two methods?

If we want to be conservative and prefer not to make change to the interface, I will move them to KafkaSystemAdmin. What do you think?

Reply:
I agree with @lindong28, should be ok to add as a default no-op method for SystemAdmin since it makes sense in the context. 

I think EventHubSystemAdmin could benefit from these lifecycle methods too. @vjagadish1989 can verify.

Reply:
I'd be +1 to @lindong28's suggestion as well. For instance, the `EventHubSystemAdmin` would benefit from doing a one-time initialization of its `EventHubClient` during `init`.

Reply:
@prateekm @vjagadish1989 my point is: the current usage pattern of creating the SystemAdmin instances is not quite following the life-cycle management of a stateful component. I did a quick grep in the code base for the calls to SystemFactory.getAdmin() calls, and there are many places:
yipan-mn1:samza_master yipan$ git grep "getAdmin" | grep -vi "test" | cut -d':' -f1 | sort -u
samza-api/src/main/java/org/apache/samza/system/SystemFactory.java
samza-azure/src/main/java/org/apache/samza/system/eventhub/EventHubSystemFactory.java
samza-core/src/main/java/org/apache/samza/config/JavaSystemConfig.java
samza-core/src/main/java/org/apache/samza/standalone/PassthroughJobCoordinator.java
samza-core/src/main/java/org/apache/samza/storage/StorageRecovery.java
samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala
samza-core/src/main/scala/org/apache/samza/coordinator/JobModelManager.scala
samza-core/src/main/scala/org/apache/samza/coordinator/stream/CoordinatorStreamSystemFactory.scala
samza-core/src/main/scala/org/apache/samza/job/JobRunner.scala
samza-core/src/main/scala/org/apache/samza/system/StreamMetadataCache.scala
samza-core/src/main/scala/org/apache/samza/system/filereader/FileReaderSystemFactory.scala
samza-elasticsearch/src/main/java/org/apache/samza/system/elasticsearch/ElasticsearchSystemFactory.java
samza-hdfs/src/main/scala/org/apache/samza/system/hdfs/HdfsSystemFactory.scala
samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManagerFactory.scala
samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemFactory.scala

For now, each invocation of SystemFactory.getAdmin() is creating a new instance of admin. If we are going to add the life-cycle methods to SystemAdmin objects, we either need to add life-cycle management code for each instance of SystemAdmin object created, or need to manage a single set of shared SystemAdmin objects (one per system) and control the life-cycle methods in a central place.